src/gui/raster/qgscolorrampshaderwidget.cpp

```diff
@@ -0,0 +1,777 @@
+  QString label;
+  int topLevelItemCount = mColormapTreeWidget->topLevelItemCount();
+  QTreeWidgetItem *currentItem = nullptr;
+  for ( int i = 0; i < topLevelItemCount; ++i )
+  {
+    currentItem = mColormapTreeWidget->topLevelItem( i );
+    //If the item is null or does not have a pixel values set, skip
+    if ( !currentItem || currentItem->text( ValueColumn ).isEmpty() )
+    {
+      continue;
+    }
+
+    if ( discrete )
+    {
+      if ( i == 0 )
+      {
+        label = "<= " + currentItem->text( ValueColumn );
+      }
+      else if ( currentItem->text( ValueColumn ).toDouble() == std::numeric_limits<double>::infinity() )
+      {
+        label = "> " + mColormapTreeWidget->topLevelItem( i - 1 )->text( ValueColumn );
+      }
+      else
+      {
+        label = mColormapTreeWidget->topLevelItem( i - 1 )->text( ValueColumn ) + " - " + currentItem->text( ValueColumn );
+      }
+    }
+    else
+    {
+      label = currentItem->text( ValueColumn );
+    }
+
+    if ( currentItem->text( LabelColumn ).startsWith( label ) )
+    {
+      allSuffixes.append( currentItem->text( LabelColumn ).mid( label.length() ) );
+    }
+  }
+  // find most common suffix
+  QStringList suffixes = QStringList( allSuffixes );
+  suffixes.removeDuplicates();
+  int max = 0;
+  QString unit;
+  for ( int i = 0; i < suffixes.count(); ++i )
+  {
+    int n = allSuffixes.count( suffixes[i] );
+    if ( n > max )
+    {
+      max = n;
+      unit = suffixes[i];
+    }
+  }
+  // Set this suffix as unit if at least used twice
+  if ( max >= 2 )
+  {
+    mUnitLineEdit->setText( unit );
+  }
+  autoLabel();
+}
+
+
+void QgsColorRampShaderWidget::mAddEntryButton_clicked()
+{
+  QgsTreeWidgetItemObject *newItem = new QgsTreeWidgetItemObject( mColormapTreeWidget );
+  newItem->setText( ValueColumn, QStringLiteral( "0" ) );
+  newItem->setBackground( ColorColumn, QBrush( QColor( Qt::magenta ) ) );
+  newItem->setText( LabelColumn, QString() );
+  newItem->setFlags( Qt::ItemIsEnabled | Qt::ItemIsEditable | Qt::ItemIsSelectable );
+  connect( newItem, &QgsTreeWidgetItemObject::itemEdited,
+           this, &QgsColorRampShaderWidget::mColormapTreeWidget_itemEdited );
+  mColormapTreeWidget->sortItems( ValueColumn, Qt::AscendingOrder );
+  autoLabel();
+
+  loadMinMaxFromTree();
+  emit widgetChanged();
+}
+
+void QgsColorRampShaderWidget::mDeleteEntryButton_clicked()
+{
+  QList<QTreeWidgetItem *> itemList;
+  itemList = mColormapTreeWidget->selectedItems();
+  if ( itemList.isEmpty() )
+  {
+    return;
+  }
+
+  Q_FOREACH ( QTreeWidgetItem *item, itemList )
+  {
+    delete item;
+  }
+
+  loadMinMaxFromTree();
+  emit widgetChanged();
+}
+
+void QgsColorRampShaderWidget::classify()
+{
+  std::unique_ptr< QgsColorRamp > ramp( btnColorRamp->colorRamp() );
+  if ( !ramp || std::isnan( mMin ) || std::isnan( mMax ) )
+  {
+    return;
+  }
+
+  QgsColorRampShader *colorRampShader = new QgsColorRampShader(
+    mMin, mMax,
+    ramp.get(),
```

Comment:
ramp.release() because it takes ownership of the color ramp

use unique_ptr for colorRampShader